README
PAR-Repository-Client
=====================

This module represents the client for PAR repositories as
implemented by the L<PAR::Repository> module.

Chances are, you should be looking at the L<PAR> module
instead. Starting with version 0.950, it supports
automatically loading any modules that aren't found on your
system from a repository. If you need finer control than that,
then this module is the right one to use.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires other modules and libraries. A complete
list of dependencies can be found in the Makefile.PL file.

COPYRIGHT AND LICENCE

Copyright (C) 2006-2008 by Steffen Mueller

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.6 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension PAR::Repository::Client.

0.24_01  Fri Feb 20 15:21 2009
    - Many, many, many improvements to the caching algorithms.
      In fact, I'd go as far as saying that this will be the first
      release where the caching works well.
    - Inter-run caching of the DBMs with local checksum
      re-calculation.
    - Introduce the private-cache-dir setting for controlling the
      caching behaviour.
    - Caching improves run-time of simple test by factor of three.
    - Enforce the use of a temporary directory as a cache dir
      in the tests much more so we are good CPAN testers citizens.

0.24  Thu Feb 19 17:16 2009
    - Promote to stable release

0.23_02  Mon Feb  2 21:14 2009
    - Attempt band-aid fix for a better error recovery in case
      of a failing dependency resolution.
    - Fix for the dependencies test (set arch and perl version
      manually)

0.23_01  Mon Feb  2 01:38 2009
    - "static_dependencies" option for automatic resolution
      of the statically declared dependencies.
    - "get_module_dependencies" and "get_script_dependencies"
      methods returning a list of all PAR distributions from
      the repository required to use a given module or script.
    - Much better abstraction of the (internal) X_dbm methods
    - Upgrading the testing repositories to include deps.
    - Require PAR 0.985 due to minor bug fix for the
      "run script from repository" use case.

0.22_01  Sun Jan 25 23:16 2009
    - Implement/fix auto-upgrading.
    - Require PAR 0.984 for upgrading.
    - Require PAR::Dist 0.43 so install_module doesn't fail
      with newer ExtUtils::Installs

0.21_03  Wed Jan 14 08:00 2009
    - Disable File::Temp exclusive locking or else BSD is
      unhappy.

0.21_02  Tue Jan 13 20:06 2009
    - Support for setting custom cache directories with
      the cache_dir parameter to new().
    - Set PAR_TEMP to a real temporary directory with cleanup
      for the tests so there are not files leftover after
      testing (RT #30018)

0.21_01  Tue Jan 13 18:00 2009
    - Fix return values of modules_dbm and scripts_dbm:
      Formerly, they would only return the hash (not the file)
      if the checksums matched.
    - Fix return value of need_dbm_update on success. Documented
      to be the empty list, was 0.
    - Fix need_dbm_update in case the checksums had never been
      fetched before.
    - Fix autovivification bug in _fetch_module(). (very minor)
    - Fix bug in _fetch_module()
    - Much, MUCH better tests.

0.21  Mon Oct 20 20:49 2008
    - Simple distribution fix. 0.20 was missing a file. Oops!

0.20  Sun Oct 19 18:20 2008
    - Require PAR 0.982: 
    --> Only load PAR.pm when really needed, otherwise load just
        the new PAR::Setup* modules to setup the environment.
    - Support for manually specifying the preferred architecture
      and perl version.

0.18  Tue Jun  3 15:29 2008
    - Don't refetch DBM files when deemed not necessary because the
      checksums match..
    - Add getters for dbm checksums in case the repository
      supports them.
    - Add PAR::Repository::Client::Util for small, common
      private utility methods.

0.17  Sun Jun  1 15:32 2008
    - No autoflushing of DBM::Deep objects.

0.16  Mon Oct 15 10:25 2007
    - Add http_timeout option for HTTP transport.
    - Minor documentation tweaks.

0.15  Mon Oct 29 13:15 2006
    - PAR::Repository::Client::prefered_distribution was silently dropping
      any_version archives because of a typo.

0.14  Mon Oct 12 16:05 2006
    - PAR::Repository::Client::HTTP::_fetch_file was generating rather
      long local file names. In fact long enough file names that some systems
      go belly-up. Now generating shorter file names.

0.13  Mon Oct  2 15:02 2006
    - Added custom installation targets.

0.12  Fri Sep 15 10:50 2006
    - Added dependency on the new PAR::Repository::Client which defines
      a unified interface for repository queries from both client and
      server side software.
    - _modules_dbm => modules_dbm
    - _scripts_dbm => scripts_dbm
    - _close_modules_dbm => close_modules_dbm
    - _close_scripts_dbm => close_scripts_dbm

0.11  Mon Aug 28 10:28 2006
    - Explicitly depending on PAR::Dist 0.18.
    - Added optional POD tests

0.10  Sat Aug 26 14:25 2006
    - Changed the repository version validation to check the
      major version and the first digit after the period only.

0.05  Tue Aug 22 16:02 2006
    - Added upgrade_module() functionality.

0.04  Tue Aug 22 12:52 2006
    - Added install_module() functionality.
    - Added "auto_install" flag to new().
    - Fixed 03prefered_distribution.t.

0.03  Fri Aug 11 16:14 2006
    - Workaround for YAML::Syck/Tiny bug.
    - Added run_script which runs a script from a repository.

0.02  Thu Aug 10 15:23 2006
    - Added text to README.
    - Added support for the repository_info.yml file.
    - Checks that a repository has a compatible version.

0.01  Tue Jul 25 14:21 2006
    - original version



t/lib/RepoMisc.pm
package RepoMisc;
use strict;
use warnings;
require Test::More;
use File::Temp ();
use File::Spec ();

Test::More::use_ok('PAR::Repository::Client');

# 3 tests
sub client_ok {
  my $path = shift;
  my $client = PAR::Repository::Client->new(
    uri => $path,
    verbosity => 3,
    cache_dir => $ENV{PAR_TEMP},
  );

  Test::More::isa_ok($client, 'PAR::Repository::Client');
  Test::More::ok(!$client->error, "no error");
  return $client;
}

# 1 test
sub set_installation_targets {
  my $client = shift;
  my $dir = File::Temp::tempdir( CLEANUP => 1 );
  unshift @main::INC, $dir;

  my %targets = (
    inst_lib => $dir,
    inst_archlib => $dir,
    inst_script => $dir,
    inst_bin => $dir,
    inst_man1dir => $dir,
    inst_man3dir => $dir,
    inst_man3dir => $dir,
    packlist_read => File::Spec->catfile($dir, '.packlist'),
    packlist_write => File::Spec->catfile($dir, '.packlist'),
  );
  $client->installation_targets(%targets);
  Test::More::is_deeply(\%targets, $client->installation_targets());

  return $dir;
}

1;


lib/PAR/Repository/Client/DBM.pm
package PAR::Repository::Client::DBM;

use 5.006;
use strict;
use warnings;

our $VERSION = '0.21';

use Carp qw/croak/;

=head1 NAME

PAR::Repository::Client::DBM - Contains all the DBM access functions

=head1 SYNOPSIS

  use PAR::Repository::Client;

=head1 DESCRIPTION

This module implements access to the underlying DBMs.

All of the methods described here shouldn't be used frivolously in user
code even if some of them are part of the API and are guaranteed not
to change.

=cut


=head2 need_dbm_update

Takes one or no arguments. Without arguments, all DBM files are
checked. With an argument, only the specified DBM file will be checked.

Returns true if either one of the following conditions match:

=over 2

=item

The repository does not support checksums.

=item

The checksums (and thus also the DBM files) haven't been
downloaded yet.

=item

The local copies of the checksums do not match those of the repository.

=back

In cases two and three above, the return value is actually the hash
reference of checksums that was fetched from the repository.

Returns the empty list if the local checksums match those of the
repository exactly.

You don't usually need to call this directly. By default, DBM files
are only fetched from the repository if necessary.

=cut

sub need_dbm_update {
  my $self = shift;
  $self->{error} = undef;

  my $check_file = shift;
  $check_file .= '.zip' if defined $check_file and not $check_file =~ /\.zip$/;

  my $support = $self->{supports_checksums};
  if (defined $support and not $support) {
    return 1;
  }

  my $checksums = $self->_dbm_checksums();

  if (not defined $checksums) {
    $self->{supports_checksums} = 0;
    return 1;
  }
  else {
    $self->{supports_checksums} = 1;
  }

  if (not defined $self->{checksums} or keys %{$self->{checksums}} == 0) {
    # never fetched checksums before.
    return $checksums;
  }
  else {
    # we fetched checksums earlier, match them
    my $local_checksums = $self->{checksums};
    if (not defined $check_file) {
      return $checksums if keys(%$local_checksums) != keys(%$checksums);
      foreach my $file (keys %$checksums) {
        return $checksums
          if not exists $local_checksums->{$file}
          or not $local_checksums->{$file} eq $checksums->{$file};
      }
    }
    else {
      return $checksums
        if not exists $local_checksums->{$check_file}
        or not exists $checksums->{$check_file} # shouldn't happen
        or not $local_checksums->{$check_file} eq $checksums->{$check_file};
    }
    return();
  }
}


=head2 modules_dbm

Fetches the C<modules_dists.dbm> database from the repository,
ties it to a L<DBM::Deep> object and returns a tied hash
reference or the empty list on failure. Second return
value is the name of the local temporary file.

In case of failure, an error message is available via
the C<error()> method.

The method uses the C<_fetch_dbm_file()> method which must be
implemented in a subclass such as L<PAR::Repository::Client::HTTP>.

=cut

sub modules_dbm {
  my $self = shift;
  return( $self->_get_a_dbm('modules', PAR::Repository::Client::MODULES_DBM_FILE()) );
}


=head2 scripts_dbm

Fetches the C<scripts_dists.dbm> database from the repository,
ties it to a L<DBM::Deep> object and returns a tied hash
reference or the empty list on failure. Second return
value is the name of the local temporary file.

In case of failure, an error message is available via
the C<error()> method.

The method uses the C<_fetch_dbm_file()> method which must be
implemented in a subclass such as L<PAR::Repository::Client::HTTP>.

=cut

sub scripts_dbm {
  my $self = shift;
  return( $self->_get_a_dbm('scripts', PAR::Repository::Client::SCRIPTS_DBM_FILE()) );
}


=head2 dependencies_dbm 

Fetches the C<dependencies.dbm> database from the repository,
ties it to a L<DBM::Deep> object and returns a tied hash
reference or the empty list on failure. Second return
value is the name of the local temporary file.

In case of failure, an error message is available via
the C<error()> method.

The method uses the C<_fetch_dbm_file()> method which must be
implemented in a subclass such as L<PAR::Repository::Client::HTTP>.

=cut

sub dependencies_dbm {
  my $self = shift;
  return( $self->_get_a_dbm('dependencies', PAR::Repository::Client::DEPENDENCIES_DBM_FILE()) );
}


=head2 close_modules_dbm

Closes the C<modules_dists.dbm> file and does all necessary
cleaning up.

This is called when the object is destroyed.

=cut

sub close_modules_dbm {
  my $self = shift;
  my $hash = $self->{modules_dbm_hash};
  return if not defined $hash;

  my $obj = tied($hash);
  $self->{modules_dbm_hash} = undef;
  undef $hash;
  undef $obj;

  unlink $self->{modules_dbm_temp_file} if not $self->{private_cache_dir};
  $self->{modules_dbm_temp_file} = undef;
  if ($self->{checksums}) {
    delete $self->{checksums}{PAR::Repository::Client::MODULES_DBM_FILE().".zip"};
  }

  return 1;
}


=head2 close_scripts_dbm

Closes the C<scripts_dists.dbm> file and does all necessary
cleaning up.

This is called when the object is destroyed.

=cut

sub close_scripts_dbm {
  my $self = shift;
  my $hash = $self->{scripts_dbm_hash};
  return if not defined $hash;

  my $obj = tied($hash);
  $self->{scripts_dbm_hash} = undef;
  undef $hash;
  undef $obj;

  unlink $self->{scripts_dbm_temp_file} if !$self->{private_cache_dir};
  $self->{scripts_dbm_temp_file} = undef;
  if ($self->{checksums}) {
    delete $self->{checksums}{PAR::Repository::Client::SCRIPTS_DBM_FILE().".zip"};
  }

  return 1;
}


=head2 close_dependencies_dbm

Closes the C<dependencies.dbm> file and does all necessary
cleaning up.

This is called when the object is destroyed.

=cut

sub close_dependencies_dbm {
  my $self = shift;
  my $hash = $self->{dependencies_dbm_hash};
  return if not defined $hash;

  my $obj = tied($hash);
  $self->{dependencies_dbm_hash} = undef;
  undef $hash;
  undef $obj;

  unlink $self->{dependencies_dbm_temp_file} if !$self->{private_cache_dir};
  $self->{dependencies_dbm_temp_file} = undef;
  if ($self->{checksums}) {
    delete $self->{checksums}{PAR::Repository::Client::DEPENDENCIES_DBM_FILE().".zip"};
  }

  return 1;
}


=head1 PRIVATE METHODS

These private methods should not be relied upon from the outside of
the module.

=head2 _get_a_dbm

This is a private method.

Generic method returning a dbm.
Requires two arguments. The type of the DBM (C<modules>,
C<scripts>, C<dependencies>), and the name of the remote
DBM file. The latter should be taken from one of the package
constants.

=cut

sub _get_a_dbm {
  my $self = shift;
  $self->{error} = undef;

  my $dbm_type       = shift;
  my $dbm_remotefile = shift;

  my $dbm_hashkey        = $dbm_type . "_dbm_hash";
  my $tempfile_hashkey   = $dbm_type . "_dbm_temp_file";
  my $dbm_remotefile_zip = $dbm_remotefile . ".zip";

  my $checksums = $self->need_dbm_update($dbm_remotefile);

  if ($self->{$dbm_hashkey}) {
    # need new dbm file?
    return($self->{$dbm_hashkey}, $self->{$tempfile_hashkey})
      if not $checksums;

    # does this particular dbm need to be updated?
    if ($self->{checksums}) {
      my $local_checksum = $self->{checksums}{$dbm_remotefile_zip};
      my $remote_checksum = $checksums->{$dbm_remotefile_zip};
      return($self->{$dbm_hashkey}, $self->{$tempfile_hashkey})
        if defined $local_checksum and defined $remote_checksum
           and $local_checksum eq $remote_checksum;
    }

    # just to make sure
    my $method = 'close_' . $dbm_type . "_dbm";
    $self->$method;
  }

  my $file;
  if ($checksums) {
    $file = $self->_fetch_dbm_file($dbm_remotefile_zip);
    # (error set by _fetch_dbm_file)
    return() if not defined $file; # or not -f $file; # <--- _fetch_dbm_file should do the stat!
  }
  else {
    # cached!
    $file = File::Spec->catfile($self->{cache_dir}, $dbm_remotefile_zip);
    $self->{error} = "Cache miss error: Expected $file to exist, but it doesn't" if not -f $file;
  }

  my ($tempfh, $tempfile) = File::Temp::tempfile(
    'temporary_dbm_XXXXX',
    UNLINK => 0,
    DIR => File::Spec->tmpdir(),
    EXLOCK => 0, # FIXME no exclusive locking or else we block on BSD. What's the right solution?
  );

  if (not $self->_unzip_file($file, $tempfile, $dbm_remotefile)) {
    $self->{error} = "Could not unzip dbm file '$file' to '$tempfile'";
    return();
  }

  $self->{$tempfile_hashkey} = $tempfile;

  my %hash;
  my $obj = tie %hash, "DBM::Deep", {
    file => $tempfile,
    locking => 1,
    autoflush => 0,
  }; 

  $self->{$dbm_hashkey} = \%hash;

  # save this dbm file checksum
  if (ref($checksums)) {
    if (not $self->{checksums}) {
      $self->{checksums} = {};
    }
    $self->{checksums}{$dbm_remotefile_zip} = $checksums->{$dbm_remotefile_zip};
  }

  return (\%hash, $tempfile);
}


=head2 _parse_dbm_checksums

This is a private method.

Given a reference to a file handle, a reference to a string
or a file name, this method parses a checksum file
and returns a hash reference associating file names
with their base64 encoded MD5 hashes.

If passed a ref to a string, the contents of the string will
be assumed to contain the checksum data.

=cut

sub _parse_dbm_checksums {
  my $self = shift;
  $self->{error} = undef;

  my $file_or_fh = shift;
  my $is_string = 0;
  my $fh;
  if (ref($file_or_fh) eq 'GLOB') {
    $fh = $file_or_fh;
  }
  elsif (ref($file_or_fh) eq 'SCALAR') {
    $is_string = 1;
  }
  else {
    open $fh, '<', $file_or_fh
      or die "Could not open file '$file_or_fh' for reading: $!";
  }

  my $hashes = {};
  my @lines;
  @lines = split /\n/, $$file_or_fh if $is_string;

  while (1) {
    local $_ = $is_string ? shift @lines : <$fh>;
    last if not defined $_;
    next if /^\s*$/ or /^\s*#/;
    my ($file, $hash) = split /\t/, $_;
    if (not defined $file or not defined $hash) {
      $self->{error} = "Error reading repository checksums.";
      return();
    }
    $hash =~ s/\s+$//;
    $hashes->{$file} = $hash;
  }

  return $hashes;
}


=head2 _calculate_cache_local_checksums

This is a private method.

Calculates the checksums of the DBMs in the local cache directory.
If the repository client isn't using a private cache directory, this
B<short circuits> and does not actually try to calculate
any checksums of potentially modified files.

Returns the checksums hash just like the checksum fetching
routine.

Maintainer note: Essentially the same code lives in
PAR::Repository's DBM code for calculating the repository checksums
in the first place.

=cut

sub _calculate_cache_local_checksums {
  my $self = shift;

  # only support inter-run cache summing if we're in a private cache dir!
  if (!$self->{private_cache_dir}) {
    return();
  }

  # find a working base64 MD5 implementation
  my $md5_function;
  eval { require Digest::MD5; $md5_function = \&Digest::MD5::md5_base64; };
  eval { require Digest::Perl::MD5;  $md5_function = \&Digest::Perl::MD5::md5_base64; } if $@;
  if ($@) {
    return();
  }
  
  my $hashes = {};
  # calculate local hashes
  foreach my $dbmfile (
      PAR::Repository::Client::MODULES_DBM_FILE(),
      PAR::Repository::Client::SCRIPTS_DBM_FILE(),
      PAR::Repository::Client::SYMLINKS_DBM_FILE(),
      PAR::Repository::Client::DEPENDENCIES_DBM_FILE(),
    ) {
    my $filepath = File::Spec->catfile($self->{cache_dir}, $dbmfile.'.zip');
    next unless -f $filepath;
    open my $fh, '<', $filepath
      or die "Could not open DBM file '$filepath' for reading: $!";
    local $/ = undef;
    my $hash = $md5_function->(<$fh>);
    close $fh;
    $hashes->{$dbmfile.'.zip'} = $hash;
  } # end foreach dbm files

  return $hashes; 
}

1;

__END__

=head1 SEE ALSO

This module is directly related to the C<PAR> project. You need to have
basic familiarity with it. Its homepage is at L<http://par.perl.org/>

See L<PAR>, L<PAR::Dist>, L<PAR::Repository>, etc.

L<PAR::Repository::Query> implements the querying interface. The methods
described in that module's documentation can be called on
C<PAR::Repository::Client> objects.

L<PAR::Repository> implements the server side creation and manipulation
of PAR repositories.

L<PAR::WebStart> is doing something similar but is otherwise unrelated.

=head1 AUTHOR

Steffen Mueller, E<lt>smueller@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright 2006-2009 by Steffen Mueller

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.6 or,
at your option, any later version of Perl 5 you may have available.

=cut


lib/PAR/Repository/Client/HTTP.pm
package PAR::Repository::Client::HTTP;

use 5.006;
use strict;
use warnings;

use vars qw/$ua/;
require LWP::Simple;
LWP::Simple->import('$ua');

use base 'PAR::Repository::Client';

use Carp qw/croak/;

our $VERSION = '0.21';

=head1 NAME

PAR::Repository::Client::HTTP - PAR repository via HTTP

=head1 SYNOPSIS

  use PAR::Repository::Client;
  
  my $client = PAR::Repository::Client->new(
    uri => 'http:///foo/repository',
    http_timeout => 20, # but default is 180s
  );

=head1 DESCRIPTION

This module implements repository accesses via HTTP.

If you create a new L<PAR::Repository::Client> object and pass it
an uri parameter which starts with C<http://> or C<https://>,
it will create an object of this class. It inherits from
C<PAR::Repository::Client>.

The repository is accessed using L<LWP::Simple>.

=head2 EXPORT

None.

=head1 METHODS

Following is a list of class and instance methods.
(Instance methods until otherwise mentioned.)

=cut


=head2 fetch_par

Fetches a .par distribution from the repository and stores it
locally. Returns the name of the local file or the empty list on
failure.

First argument must be the distribution name to fetch.

=cut

sub fetch_par {
  my $self = shift;
  $self->{error} = undef;
  my $dist = shift;
  return() if not defined $dist;

  my $url = $self->{uri};
  $url =~ s/\/$//;

  my ($n, $v, $a, $p) = PAR::Dist::parse_dist_name($dist);
  $url .= "/$a/$p/$n-$v-$a-$p.par";

  my $file = $self->_fetch_file($url);

  if (not defined $file) {
    $self->{error} = "Could not fetch distribution from URI '$url'";
    return();
  }

  return $file;
}


{
  my %escapes;
  sub _fetch_file {
    my $self = shift;
    $self->{error} = undef;
    my $file = shift;

    my $cache_dir = $self->{cache_dir}; # used to be PAR_TEMP, but now configurable
    %escapes = map { chr($_) => sprintf("%%%02X", $_) } 0..255 unless %escapes;

    $file =~ m!/([^/]+)$!;
    my $local_file = (defined($1) ? $1 : $file);
    $local_file =~ s/([^\w\._])/$escapes{$1}/g;
    $local_file = File::Spec->catfile( $self->{cache_dir}, $local_file );

    my $timeout = $self->{http_timeout};
    my $old_timeout = $ua->timeout();
    $ua->timeout($timeout) if defined $timeout;
    my $rc = LWP::Simple::mirror( $file, $local_file );
    $ua->timeout($old_timeout) if defined $timeout;
    if (!LWP::Simple::is_success($rc) and not $rc == HTTP::Status::RC_NOT_MODIFIED()) {
      $self->{error} = "Error $rc: " . LWP::Simple::status_message($rc) . " ($file)\n";
      return();
    }

    return $local_file if -f $local_file;
    return();
  }
}


sub _fetch_as_data {
  my $self = shift;
  $self->{error} = undef;
  my $file = shift;

  my $timeout = $self->{http_timeout};
  my $old_timeout = $ua->timeout();
  $ua->timeout($timeout) if defined $timeout;
  my $data = LWP::Simple::get( $file );
  $ua->timeout($old_timeout) if defined $timeout;

  return $data if defined $data;

  $self->{error} = "Could not get '$file' from repository";
  return();
}


=head2 validate_repository

Makes sure the repository is valid. Returns the empty list
if that is not so and a true value if the repository is valid.

Checks that the repository version is compatible.

The error message is available as C<$client->error()> on
failure.

=cut

sub validate_repository {
  my $self = shift;
  $self->{error} = undef;

  my $mod_db = $self->modules_dbm;
  return() if not defined $mod_db;

  return() if not $self->validate_repository_version;

  return 1;
}


=head2 _repository_info

Returns a YAML::Tiny object representing the repository meta
information.

This is a private method.

=cut

sub _repository_info {
  my $self = shift;
  $self->{error} = undef;
  return $self->{info} if defined $self->{info};

  my $url = $self->{uri};
  $url =~ s/\/$//;

  my $file = $self->_fetch_file(
    $url.'/'.PAR::Repository::Client::REPOSITORY_INFO_FILE()
  );

  return() if not defined $file;

  my $yaml = YAML::Tiny->new->read($file);
  if (not defined $yaml) {
    $self->{error} = "Error reading repository info from YAML file.";
    return();
  }

  # workaround for possible YAML::Syck/YAML::Tiny bug
  # This is not the right way to do it!
  @$yaml = ($yaml->[1]) if @$yaml > 1;

  $self->{info} = $yaml;
  return $yaml;
}


=head2 _fetch_dbm_file

This is a private method.

Fetches a dbm (index) file from the repository and
returns the name of the temporary local file or the
empty list on failure.

An error message is available via the C<error()>
method in case of failure.

=cut

sub _fetch_dbm_file {
  my $self = shift;
  $self->{error} = undef;
  my $file = shift;
  return if not defined $file;

  my $url = $self->{uri};
  $url =~ s/\/$//;

  my $local = $self->_fetch_file("$url/$file");
  return() if not defined $local or not -f $local;

  return $local;
}


=head2 _dbm_checksums

This is a private method.

If the repository has a checksums file (new feature of
C<PAR::Repository> 0.15), this method returns a hash  
associating the DBM file names (e.g. C<foo_bar.dbm.zip>)
with their MD5 hashes (base 64).

This method B<always> queries the repository and never caches
the information locally. That's the whole point of having the
checksums.

In case the repository does not have checksums, this method
returns the empty list, so check the return value!
The error message (see the C<error()> method) will be
I<"Repository does not support checksums"> in that case.

=cut

sub _dbm_checksums {
  my $self = shift;
  $self->{error} = undef;

  my $url = $self->{uri};
  $url =~ s/\/$//;

  my $data = $self->_fetch_as_data(
    $url.'/'.PAR::Repository::Client::DBM_CHECKSUMS_FILE()
  );

  if (not defined $data) {
    $self->{error} = "Repository does not support checksums";
    return();
  }

  return $self->_parse_dbm_checksums(\$data);
}


=head2 _init

This private method is called by the C<new()> method of
L<PAR::Repository::Client>. It is used to initialize
the client object and C<new()> passes it a hash ref to
its arguments.

Should return a true value on success.

=cut

sub _init {
  my $self = shift;
  my $args = shift || {};
  # We implement additional object attributes here
  $self->{http_timeout} = $args->{http_timeout};
  $self->{http_timeout} = 180 if not defined $self->{http_timeout};

  return 1;
}



1;
__END__

=head1 SEE ALSO

This module is part of the L<PAR::Repository::Client> distribution.

This module is directly related to the C<PAR> project. You need to have
basic familiarity with it. The PAR homepage is at L<http://par.perl.org>.

See L<PAR>, L<PAR::Dist>, L<PAR::Repository>, etc.

L<PAR::Repository> implements the server side creation and manipulation
of PAR repositories.

L<PAR::WebStart> is doing something similar but is otherwise unrelated.

The repository access is done via L<LWP::Simple>.

=head1 AUTHOR

Steffen Mueller, E<lt>smueller@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2006-2009 by Steffen Mueller

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.6 or,
at your option, any later version of Perl 5 you may have available.

=cut


lib/PAR/Repository/Client/Util.pm
package PAR::Repository::Client::Util;

use 5.006;
use strict;
use warnings;

our $VERSION = '0.23';

use Carp qw/croak/;

=head1 NAME

PAR::Repository::Client::Util - Small helper methods common to all implementations

=head1 SYNOPSIS

  use PAR::Repository::Client;

=head1 DESCRIPTION

This module implements small helper methods which are common to all
L<PAR::Repository::Client> implementations.

=head1 PRIVATE METHODS

These private methods should not be relied upon from the outside of
the module.

=head2 _unzip_file

This is a private method. Callable as class or instance method.

Unzips the file given as first argument to the file
given as second argument.
If a third argument is used, the zip member of that name
is extracted. If the zip member name is omitted, it is
set to the target file name.

Returns the name of the unzipped file.

=cut

sub _unzip_file {
  my $class = shift;
  my $file = shift;
  my $target = shift;
  my $member = shift;
  $member = $target if not defined $member;
  return unless -f $file;

  my $zip = Archive::Zip->new;
  local %SIG;
  $SIG{__WARN__} = sub { print STDERR $_[0] unless $_[0] =~ /\bstat\b/ };

  return unless $zip->read($file) == Archive::Zip::AZ_OK()
    and $zip->extractMember($member, $target) == Archive::Zip::AZ_OK();

  return $target;
}


# given a distribution name, recursively determines all distributions
# it depends on
sub _resolve_static_dependencies {
  my $self = shift;
  my $distribution = shift;

  my ($deph) = $self->dependencies_dbm();
  return([]) if not exists $deph->{$distribution};
  
  my ($modh) = $self->modules_dbm();

  my @module_queue = (keys %{$deph->{$distribution}});
  my @dep_dists;
  my %module_seen;
  my %dist_seen;

  while (@module_queue) {
    #use Data::Dumper; warn Dumper \@module_queue;
    my $module = shift @module_queue;
    next if $module_seen{$module}++;
    next if not exists $modh->{$module}; # FIXME should this be somehow reported?
    my $dist = $self->prefered_distribution($module, $modh->{$module});
    next if not defined $dist;
    next if $dist_seen{$dist}++;
    push @dep_dists, $dist;
    push @module_queue, keys %{$deph->{$dist}} if exists $deph->{$dist};
  }

  return \@dep_dists;
}

sub generate_private_cache_dir {
  my $self = shift;
  my $uri = $self->{uri};
  my $digester = PAR::SetupTemp::_get_digester(); # requires PAR 0.987!
  $digester->add($uri);
  my $digest = $digester->b64digest();
  $digest =~ s/\W/_/g;
  my $user_temp_dir = PAR::SetupTemp::_get_par_user_tempdir();
  my $priv_cache_dir = File::Spec->catdir($user_temp_dir, "par-repo-$digest");
  return $priv_cache_dir;
}

1;
__END__

=head1 SEE ALSO

This module is directly related to the C<PAR> project. You need to have
basic familiarity with it. Its homepage is at L<http://par.perl.org/>

See L<PAR>, L<PAR::Dist>, L<PAR::Repository>, etc.

L<PAR::Repository::Query> implements the querying interface. The methods
described in that module's documentation can be called on
C<PAR::Repository::Client> objects.

L<PAR::Repository> implements the server side creation and manipulation
of PAR repositories.

L<PAR::WebStart> is doing something similar but is otherwise unrelated.

=head1 AUTHOR

Steffen Mueller, E<lt>smueller@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright 2006-2009 by Steffen Mueller

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.6 or,
at your option, any later version of Perl 5 you may have available.

=cut


Makefile.PL
use 5.006;
use ExtUtils::MakeMaker;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.

WriteMakefile(
    NAME              => 'PAR::Repository::Client',
    VERSION_FROM      => 'lib/PAR/Repository/Client.pm', # finds $VERSION
    LICENSE => 'perl',
    PREREQ_PM         => {
#'perl'      => '5.6',
        'DBM::Deep' => '1.0012',
        'version' => '0.50',
        'File::Spec' => '0',
        'PAR::Dist' => '0.44',
        'PAR' => '0.985',
        'LWP::Simple' => '0',
        'Archive::Zip' => '0',
        'YAML::Tiny' => '0',
        'PAR::Repository::Query' => '0.12',
    }, # e.g., Module::Name => 1.1
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM  => 'lib/PAR/Repository/Client.pm', # retrieve abstract from module
       AUTHOR         => 'Steffen Müller <smueller@cpan.org>') : ()),
);
